adapters/taboola/taboola.go

```diff
@@ -89,8 +117,50 @@ func (a *adapter) MakeBids(request *openrtb2.BidRequest, requestData *adapters.R
 	return bidResponse, errs
 }
 
-func createTaboolaRequest(request *openrtb2.BidRequest) (taboolaRequest *openrtb2.BidRequest, errors []error) {
+func (a *adapter) buildRequest(request *openrtb2.BidRequest) (*adapters.RequestData, error) {
+	requestJSON, err := json.Marshal(request)
+	if err != nil {
+		return nil, err
+	}
+
+	//set MediaType based on first imp
+	var mediaType string
+	if request.Imp[0].Banner != nil {
+		mediaType = DISPLAY_ENDPOINT_PREFIX
+	} else if request.Imp[0].Native != nil {
+		mediaType = NATIVE_ENDPOINT_PREFIX
+	} else {
+		return nil, fmt.Errorf("unsupported media type for imp: %v", request.Imp[0])
+	}
+
+	url, err := a.buildEndpointURL(request.Site.ID, mediaType)
+	if err != nil {
+		return nil, err
+	}
+
+	requestData := &adapters.RequestData{
+		Method: "POST",
+		Uri:    url,
+		Body:   requestJSON,
+	}
+
+	return requestData, nil
+}
+
+// Builds endpoint url based on adapter-specific pub settings from imp.ext
+func (a *adapter) buildEndpointURL(publisherId string, mediaType string) (string, error) {
```

Comment:
`adapter` need to suffice `adapters.Bidder` interface. Therefore methods like MakeBids, MakeRequests are defined on adapter.

However there is no such need for buildEndpointURL. So buildEndpointURL doesn't have to be defined on `(a *adapter)`

```
func  buildEndpointURL(publisherId string, mediaType string) (string, error) {
...
}
```

Reply:
@onkarvhanumante  did the changes as you suggested. there are no longer an adapter methods.

Reply:
May I reopen this discussion please? This change is not needed, refer to `buildEndpointURL` in other adapters, we expect this function to get a receiver.

Reply:
`buildEndpointURL` can exist as a function receiver in this case to access the endpoint template.

Reply:
I appreciate each feedback you provide, we also aim towards a more clean efficient code and to obey Prebid convention. 
But this is going back and forth for a while, If you could write your all agree on changes that need to be done so once and for all we will do the require changes will be really great. 

this PR https://github.com/prebid/prebid-server/pull/2522 was on hold  till this one is merge, because I didn't think this PR will take this significant amount of time, so I just merged it to this PR currently. 

@SyntaxNode 
@VeronikaSolovei9 
@onkarvhanumante

Reply:
I'm sorry about this. 
Please revert it to original `func (a *adapter) buildEndpointURL(publisherId string, mediaType string)`
I think general suggestion here is to use receiver functions where you need to access adapter fields.

Reply:
@VeronikaSolovei9  done :)